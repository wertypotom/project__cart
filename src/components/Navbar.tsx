import React, { useContext, useState } from 'react';
import {
  Navbar as NavbarBS,
  Container,
  Nav,
  Button,
  Offcanvas,
} from 'react-bootstrap';
import { NavLink } from 'react-router-dom';
import CartContext from '../context/cartContext';
import CartPicture from './CartPicture';
import CartPictureContainer from './CartCard/CartPictureContainer';

function Navbar() {
  const { items, toggleItem } = useContext(CartContext);
  const [show, setShow] = useState(false);
  const totalItems = items.reduce((r, i) => (r += i.amount), 0);

  return (
    <>
      <NavbarBS sticky="top" className="bg-white shadow-sm mb-3">
        <Container>
          <Nav className="me-auto">
            <Nav.Link to="/" as={NavLink}>
              Home
            </Nav.Link>
            <Nav.Link to="/store" as={NavLink}>
              Store
            </Nav.Link>
            <Nav.Link to="/about" as={NavLink}>
              About
            </Nav.Link>
          </Nav>
          <Button
            style={{ width: '3rem', height: '3rem', position: 'relative' }}
            variant="outline-primary"
            className="rounded-circle"
            onClick={() => setShow(true)}
          >
            <CartPicture />
            <div
              className="rounded-circle bg-danger d-flex justify-content-center align-items-center"
              style={{
                color: 'white',
                width: '1.5em',
                height: '1.5em',
                position: 'absolute',
                bottom: -5,
                right: -10,
              }}
            >
              {totalItems}
            </div>
          </Button>
        </Container>
      </NavbarBS>
      <Offcanvas show={show} onHide={() => setShow(false)} placement="end">
        <Offcanvas.Header closeButton>
          <Offcanvas.Title>Cart</Offcanvas.Title>
        </Offcanvas.Header>
        <Offcanvas.Body>
          <CartPictureContainer items={items} toggleItem={toggleItem} />
        </Offcanvas.Body>
      </Offcanvas>
    </>
  );
}

export default Navbar;
