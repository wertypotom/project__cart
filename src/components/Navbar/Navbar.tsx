import React, { useContext, useState } from 'react';
import {
  Navbar as NavbarBS,
  Container,
  Nav,
  Button,
  Offcanvas,
} from 'react-bootstrap';
import { NavLink } from 'react-router-dom';
import CartContext from '../../context/cartContext';
import CartItemsContainer from '../CartCard/CartItemsContainer';
import CartPicture from '../CartPicture';

function Navbar() {
  const { items } = useContext(CartContext);
  const [show, setShow] = useState(false);
  const totalItems = items.reduce((r, i) => (r += i.amount), 0);

  return (
    <>
      <NavbarBS sticky="top" className="bg-white shadow-sm mb-3">
        <Container>
          <Nav className="me-auto">
            <Nav.Link data-testid="link-home" to="/" as={NavLink}>
              Home
            </Nav.Link>
            <Nav.Link data-testid="link-store" to="/store" as={NavLink}>
              Store
            </Nav.Link>
            <Nav.Link data-testid="link-about" to="/about" as={NavLink}>
              About
            </Nav.Link>
          </Nav>
          <Button
            style={{ width: '3rem', height: '3rem', position: 'relative' }}
            variant="outline-primary"
            className="rounded-circle"
            onClick={() => setShow(true)}
          >
            <CartPicture />
            <div
              className="rounded-circle bg-danger d-flex justify-content-center align-items-center"
              style={{
                color: 'white',
                width: '1.5em',
                height: '1.5em',
                position: 'absolute',
                bottom: -5,
                right: -10,
              }}
            >
              {totalItems}
            </div>
          </Button>
        </Container>
      </NavbarBS>
      <Offcanvas show={show} onHide={() => setShow(false)} placement="end">
        <Offcanvas.Header closeButton>
          <Offcanvas.Title>Cart</Offcanvas.Title>
        </Offcanvas.Header>
        <Offcanvas.Body>
          <CartItemsContainer />
        </Offcanvas.Body>
      </Offcanvas>
    </>
  );
}

export default Navbar;
