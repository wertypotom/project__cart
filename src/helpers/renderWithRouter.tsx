import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import AppRouter from '../components/AppRouter/AppRouter';

const renderWithRouter = (
  component: React.ReactElement | null,
  initialPath = '/'
) => {
  return (
    <MemoryRouter initialEntries={[initialPath]}>
      {component}
      <AppRouter />
    </MemoryRouter>
  );
};

export default renderWithRouter;
