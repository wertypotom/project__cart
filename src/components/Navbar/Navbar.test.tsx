import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { describe, it, expect, vi } from 'vitest';
import renderWithRouter from '../../helpers/renderWithRouter';
import Navbar from './Navbar';

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: vi.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: vi.fn(), // устарело
    removeListener: vi.fn(), // устарело
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
    dispatchEvent: vi.fn(),
  })),
});

describe('Testind Navbar Component', () => {
  it('Should correctly render Navbar', () => {
    const component = render(renderWithRouter(<Navbar />));
    expect(component).toMatchSnapshot();
  });

  it('Should open cart on click', () => {
    render(renderWithRouter(<Navbar />));

    const toggleBtn = screen.getByRole('button');
    expect(toggleBtn).toBeInTheDocument();

    fireEvent.click(toggleBtn);

    const totalCount = screen.getByText(/Cart/i);
    expect(totalCount).toBeInTheDocument();
  });
});
