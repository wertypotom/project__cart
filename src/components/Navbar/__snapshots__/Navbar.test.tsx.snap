// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testind Navbar Component > Should correctly render Navbar 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="bg-white shadow-sm mb-3 navbar navbar-expand navbar-light sticky-top"
      >
        <div
          class="container"
        >
          <div
            class="me-auto navbar-nav"
          >
            <a
              aria-current="page"
              class="nav-link active"
              data-testid="link-home"
              href="/"
            >
              Home
            </a>
            <a
              class="nav-link"
              data-testid="link-store"
              href="/store"
            >
              Store
            </a>
            <a
              class="nav-link"
              data-testid="link-about"
              href="/about"
            >
              About
            </a>
          </div>
          <button
            class="rounded-circle btn btn-outline-primary"
            style="width: 3rem; height: 3rem; position: relative;"
            type="button"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2.25 3h1.386c.51 0 .955.343 1.087.835l.383 1.437M7.5 14.25a3 3 0 00-3 3h15.75m-12.75-3h11.218c1.121-2.3 2.1-4.684 2.924-7.138a60.114 60.114 0 00-16.536-1.84M7.5 14.25L5.106 5.272M6 20.25a.75.75 0 11-1.5 0 .75.75 0 011.5 0zm12.75 0a.75.75 0 11-1.5 0 .75.75 0 011.5 0z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            <div
              class="rounded-circle bg-danger d-flex justify-content-center align-items-center"
              style="color: white; width: 1.5em; height: 1.5em; position: absolute; bottom: -5px; right: -10px;"
            >
              0
            </div>
          </button>
        </div>
      </nav>
      <div>
        Home page
      </div>
    </div>
  </body>,
  "container": <div>
    <nav
      class="bg-white shadow-sm mb-3 navbar navbar-expand navbar-light sticky-top"
    >
      <div
        class="container"
      >
        <div
          class="me-auto navbar-nav"
        >
          <a
            aria-current="page"
            class="nav-link active"
            data-testid="link-home"
            href="/"
          >
            Home
          </a>
          <a
            class="nav-link"
            data-testid="link-store"
            href="/store"
          >
            Store
          </a>
          <a
            class="nav-link"
            data-testid="link-about"
            href="/about"
          >
            About
          </a>
        </div>
        <button
          class="rounded-circle btn btn-outline-primary"
          style="width: 3rem; height: 3rem; position: relative;"
          type="button"
        >
          <svg
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.25 3h1.386c.51 0 .955.343 1.087.835l.383 1.437M7.5 14.25a3 3 0 00-3 3h15.75m-12.75-3h11.218c1.121-2.3 2.1-4.684 2.924-7.138a60.114 60.114 0 00-16.536-1.84M7.5 14.25L5.106 5.272M6 20.25a.75.75 0 11-1.5 0 .75.75 0 011.5 0zm12.75 0a.75.75 0 11-1.5 0 .75.75 0 011.5 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
          <div
            class="rounded-circle bg-danger d-flex justify-content-center align-items-center"
            style="color: white; width: 1.5em; height: 1.5em; position: absolute; bottom: -5px; right: -10px;"
          >
            0
          </div>
        </button>
      </div>
    </nav>
    <div>
      Home page
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
